### PR TITLE
Updating project to use Poetry as build system and add some automated checks

### PR DESCRIPTION
**Pull request by liz-is**
~~Draft for now, this PR is to see if Github Actions will work...~~

- Convert pyproject.toml to use Poetry to build the project and manage dependencies
- Add Sphinx and sphinx-autoapi to build docs
- Add Github actions to automatically run tests, linting with Ruff, and type checking with Pyright
- Add ReadTheDocs integration to build and host docs on PRs/main branch

Not yet implemented:
- Code format checking with Ruff
- The docs build, but with warnings - I will fix these later, need to do some digging to understand why they occur first

### CODE REVIEW

File: .github/workflows/python-package.yml [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+        poetry-version: ["1.8.3"]
+
+    permissions:
+      pull-requests: write
+    
+    steps:
+    - uses: actions/checkout@v4
+    
+    - name: "Setup Python, Poetry and Dependencies"
+      uses: packetcoders/action-setup-cache-python-poetry@main
```
Comment (LouiseABowler) [inline .github/workflows/python-package.yml:25]: I'm a bit wary of this being `@main`, as it means any (potentially large) project updates could be pulled in without warning. Is there a disadvantage to [cacheing using the standard `setup-python`](https://github.com/actions/setup-python/blob/main/docs/advanced-usage.md#caching-packages) action? (Assuming we're only planning to test against the one poetry version, which I think would be fine.)

File: docs/conf.py [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+# -- Options for HTML output -------------------------------------------------
+# https://www.sphinx-doc.org/en/master/usage/configuration.html#options-for-html-output
+
+html_theme = 'alabaster'
+html_static_path = ['_static']
+
+# -- AutoAPI -----------------------------------------------------------------
+# https://sphinx-autoapi.readthedocs.io/en/latest/
+autoapi_dirs = ['../rex_ai']
+autodoc_typehints = 'description'
```
Comment (LouiseABowler) [inline docs/conf.py:35]:
```suggestion
autodoc_typehints = 'description'

```
Tiny fix for lack of newline at end of file.

File: notes/notes.norg
Comment (LouiseABowler) [inline notes/notes.norg:1]: Should the contents of this file be captured elsewhere, e.g. in issues for future addition into docs?

File: pyproject.toml [hunk 3]
```diff
@@ -1,73 +1,47 @@
-[project]
-name = "ReX-AI"
+[tool.poetry]
+name = "rex-ai"
```
Comment (LouiseABowler) [inline pyproject.toml:2]: I noticed that the docs build URL uses "rex-xai" - can/should we keep to one or the other for consistency?

**Review by LouiseABowler — COMMENTED** [review]
This is looking really good @liz-is 😄 I know you're still working on the PR so I can take another look when you're done if you like, but here are a handful of small changes/questions for now.

File: .github/workflows/python-package.yml (hunk 1 above)
Comment (liz-is) [reply .github/workflows/python-package.yml:25]: I had come across issues [like this](https://github.com/actions/setup-python/issues/846), which made me unsure about using that approach (the issue mentioned there is the one noted under the poetry section in your link as well). But I do think that issue is unlikely to be relevant for us, so we can switch, or I could alternatively pin this action at the current version 1.2.

File: pyproject.toml (hunk 3 above)
Comment (liz-is) [reply pyproject.toml:2]: @kellino what do you think about the naming - should we switch to use "rex-xai" everywhere?

Comment (kellino) [reply pyproject.toml:2]: Yes, that's the most logical approach @liz-is. ReX-AI would imply that we are an AI framework of some sort, whereas we are an XAI (explainable AI) framework.

File: scripts/pytorch.py
[the remark this answers is not in the document]
Comment (kellino) [reply scripts/pytorch.py:1]: @liz-is  the scripts directory should probably be excluded from github-actions. `torchvision` is not a dependency of the project, nor will it ever be as far as I can tell. This is just an example of how one might use it. At least we should silence these errors

File: .github/workflows/python-package.yml [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+        python-version: ["3.12"]
+        poetry-version: ["1.8.3"]
+
+    permissions:
+      pull-requests: write
+    
+    steps:
+    - uses: actions/checkout@v4
+    
+    - name: "Setup Python, Poetry and Dependencies"
```
Comment (LouiseABowler) [inline .github/workflows/python-package.yml:24]:
```suggestion
    - name: Setup Python, Poetry and Dependencies
```
Can skip quotation marks here to be consistent with other sections

File: .github/workflows/python-package.yml [hunk 5]
```diff
@@ -0,0 +1,52 @@
+# This workflow will install Python dependencies, run tests and lint with a variety of Python versions
+# For more information see: https://docs.github.com/en/actions/automating-builds-and-tests/building-and-testing-python
+
+name: Python package
```
Comment (LouiseABowler) [inline .github/workflows/python-package.yml:4]: Could this title be more descriptive about what's happening in this workflow? Maybe "Run tests and checks"?

File: scripts/pytorch.py
[the remark this answers is not in the document]
Comment (kellino) [reply scripts/pytorch.py:1]: Yes, I like this suggestion. Less noise is always good.

**Pull request merged**